### PR TITLE
FINERACT-1841 Fix Duplicated journal entries when a chargeback happens

### PR DESCRIPTION
**Pull request by josehernandezfintecheandomx**
## Description

After the Chargeback transaction the Disburse and Repayment journal entries were being re-created

[FINERACT-1841](https://issues.apache.org/jira/browse/FINERACT-1841)

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/interoperation/service/InteropServiceImpl.java [hunk 1]
```diff
@@ -354,7 +354,7 @@ public InteropQuoteResponseData createQuote(@NotNull JsonCommand command) {
 
         final BigDecimal fee;
         if (transactionType.isDebit()) {
-            fee = savingsAccount.calculateWithdrawalFee(request.getAmount().getAmount());
+            fee = savingsAccount.calculateTransactionFee(request.getAmount().getAmount());
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/interoperation/service/InteropServiceImpl.java]: This must be an unrelated change, isn't it? If so, let's get rid of it.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/domain/Charge.java [hunk 2]
```diff
@@ -223,12 +223,12 @@ private Charge(final String name, final BigDecimal amount, final String currency
                         .failWithCodeNoParameterAddedToErrorCode("not.allowed.charge.calculation.type.for.savings");
             }
 
-            if (!(ChargeTimeType.fromInt(getChargeTimeType()).isWithdrawalFee()
-                    || ChargeTimeType.fromInt(getChargeTimeType()).isSavingsNoActivityFee())
+            final ChargeTimeType chargeTimeType = ChargeTimeType.fromInt(getChargeTimeType());
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/domain/Charge.java]: This is also unrelated.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/domain/ChargeTimeType.java [hunk 3]
```diff
@@ -37,7 +37,8 @@ public enum ChargeTimeType {
     SHAREACCOUNT_ACTIVATION(13, "chargeTimeType.activation"), // only for loan
     SHARE_PURCHASE(14, "chargeTimeType.sharespurchase"), SHARE_REDEEM(15, "chargeTimeType.sharesredeem"),
 
-    SAVINGS_NOACTIVITY_FEE(16, "chargeTimeType.savingsNoActivityFee");
+    SAVINGS_NOACTIVITY_FEE(16, "chargeTimeType.savingsNoActivityFee"), // only for savings
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/domain/ChargeTimeType.java]: This is also unrelated.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/service/ChargeDropdownReadPlatformServiceImpl.java [hunk 4]
```diff
@@ -102,7 +102,8 @@ public List<EnumOptionData> retrieveSavingsCollectionTimeTypes() {
                 // chargeTimeType(ChargeTimeType.SAVINGS_CLOSURE),
                 chargeTimeType(ChargeTimeType.WITHDRAWAL_FEE), chargeTimeType(ChargeTimeType.ANNUAL_FEE),
                 chargeTimeType(ChargeTimeType.MONTHLY_FEE), chargeTimeType(ChargeTimeType.WEEKLY_FEE),
-                chargeTimeType(ChargeTimeType.OVERDRAFT_FEE), chargeTimeType(ChargeTimeType.SAVINGS_NOACTIVITY_FEE));
+                chargeTimeType(ChargeTimeType.OVERDRAFT_FEE), chargeTimeType(ChargeTimeType.SAVINGS_NOACTIVITY_FEE),
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/service/ChargeDropdownReadPlatformServiceImpl.java]: This is also unrelated.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/service/ChargeEnumerations.java [hunk 5]
```diff
@@ -58,6 +58,10 @@ public static EnumOptionData chargeTimeType(final ChargeTimeType type) {
                 optionData = new EnumOptionData(ChargeTimeType.WITHDRAWAL_FEE.getValue().longValue(),
                         ChargeTimeType.WITHDRAWAL_FEE.getCode(), "Withdrawal Fee");
             break;
+            case DEPOSIT_FEE:
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/service/ChargeEnumerations.java]: This is also unrelated.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/service/ChargeWritePlatformServiceJpaRepositoryImpl.java [hunk 6]
```diff
@@ -287,7 +287,7 @@ private boolean isAnyLoanProductsAssociateWithThisCharge(final Long chargeId) {
     }
 
     private boolean isAnySavingsProductsAssociateWithThisCharge(final Long chargeId) {
-        final String sql = "select (CASE WHEN (exists (select 1 from m_savings_product_charge sc where sc.charge_id = ?)) = 1 THEN 'true' ELSE 'false' END)";
+        final String sql = "select (CASE WHEN exists (select 1 from m_savings_product_charge sc where sc.charge_id = ?) THEN 'true' ELSE 'false' END)";
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/service/ChargeWritePlatformServiceJpaRepositoryImpl.java]: This is also unrelated.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/data/SavingsAccountData.java [hunk 7]
```diff
@@ -901,7 +901,7 @@ private SavingsAccountData(final Long id, final String accountNo, final EnumOpti
         this.savingsAmountOnHold = savingsAmountOnHold;
     }
 
-    private SavingsAccountChargeData calculateWithdrawalFee() {
+    private SavingsAccountChargeData calculateTransactionFee() {
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/data/SavingsAccountData.java]: This is also unrelated.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccount.java [hunk 8]
```diff
@@ -1150,6 +1150,9 @@ public SavingsAccountTransaction deposit(final SavingsAccountTransactionDTO tran
             addTransaction(transaction);
         }
 
+        payDepositFee(transactionDTO.getTransactionAmount(), transactionDTO.getTransactionDate(), transactionDTO.getAppUser(),
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccount.java]: This is also unrelated.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccount.java [hunk 9]
```diff
@@ -1347,6 +1350,41 @@ private void payWithdrawalFee(final BigDecimal transactionAmount, final LocalDat
         }
     }
 
+    private void payDepositFee(final BigDecimal transactionAmount, final LocalDate transactionDate, final AppUser user,
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccount.java]: This is also unrelated.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccountCharge.java [hunk 10]
```diff
@@ -720,6 +720,10 @@ public boolean isWithdrawalFee() {
         return ChargeTimeType.fromInt(this.chargeTime).isWithdrawalFee();
     }
 
+    public boolean isDepositFee() {
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccountCharge.java]: This is also unrelated.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java [hunk 11]
```diff
@@ -6215,7 +6215,8 @@ private ChangedTransactionDetail handleRefundTransaction(final LoanTransaction l
     }
 
     public void handleChargebackTransaction(final LoanTransaction chargebackTransaction,
-            final LoanLifecycleStateMachine loanLifecycleStateMachine) {
+            final LoanLifecycleStateMachine loanLifecycleStateMachine, final List<Long> existingTransactionIds,
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: I don't really like this change because it'll bring side-effects to an already complex codebase.

Why can't we just do the findExistingTransactionIds and findExistingReversedTransactionIds where this is being called?

**Review by galovics — CHANGES_REQUESTED** [review]

File: fineract-provider/src/main/java/org/apache/fineract/interoperation/service/InteropServiceImpl.java (hunk 1 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/interoperation/service/InteropServiceImpl.java]: Done removed!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/domain/Charge.java (hunk 2 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/domain/Charge.java]: Done removed!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/domain/ChargeTimeType.java (hunk 3 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/domain/ChargeTimeType.java]: Done removed!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/service/ChargeDropdownReadPlatformServiceImpl.java (hunk 4 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/service/ChargeDropdownReadPlatformServiceImpl.java]: Done removed!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/data/SavingsAccountData.java (hunk 7 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/data/SavingsAccountData.java]: Done removed!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccount.java (hunk 8 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccount.java]: Done removed!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccount.java (hunk 9 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccount.java]: Done removed!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccountCharge.java (hunk 10 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccountCharge.java]: Done removed!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java (hunk 11 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: To be honest It's the same, I mean, It's the unique part of the code that uses this.

Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: Done, updated

**Comment by galovics** [comment]
@josehernandezfintecheandomx I still see changes related to SavingsAccount, can you please undo those? As soon as that's done, we can merge it.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/service/ChargeEnumerations.java (hunk 5 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/service/ChargeEnumerations.java]: Removed

**Comment by josehernandezfintecheandomx** [comment]
> @josehernandezfintecheandomx I still see changes related to SavingsAccount, can you please undo those? As soon as that's done, we can merge it.

@galovics The savings accounts changes were removed

**Comment by galovics** [comment]
@josehernandezfintecheandomx Still seeing it:
image

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/service/ChargeWritePlatformServiceJpaRepositoryImpl.java (hunk 6 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/service/ChargeWritePlatformServiceJpaRepositoryImpl.java]: This is a bug but anyway It was restored

**Review by galovics — APPROVED** [review]
